Wikipedia/Code/WMFExploreSectionControllerCache.m

```diff
@@ -43,22 +50,80 @@ - (instancetype)initWithDataStore:(MWKDataStore*)dataStore {
         self.dataStore                              = dataStore;
         self.sectionControllersBySection            = [[NSCache alloc] init];
         self.sectionControllersBySection.countLimit = [WMFExploreSection totalMaxNumberOfSections];
-        self.sectionsBySectionController            = [NSMapTable mapTableWithKeyOptions:NSMapTableWeakMemory | NSMapTableObjectPointerPersonality
-                                                                            valueOptions:NSMapTableWeakMemory];
+
+        /*
+         Controllers (keys in the map) are retained by the NSCache and can be evicted at any time. On the other hand,
+         sections (values in the map) are "value types" which should be owned by the receiver.  In other words, sections
+         are created as needed, and not retained for any particular reason, and it would be unintentional for a value in
```

Comment:
>This is mostly pedantic, 

Not a good start :)

> since the cache retains (note, not copies) the sections as its keys, but I should be able to retrieve a controller for a copy of the section of the section I previously associated it with. 

Not following. Retrieving a controller for a copy of a section is done by the cache, not by this object. Did you say this backwards?

> IOW, it shouldn't matter that it's a copy of the original key as long as they're equivalent. 


This seems like a separate issue from retaining. This has to do with the object personality setting, no?


> On the other hand, I can't think of any cases—pedantic or practical—that I would want sections to be released in the reverse lookup map. 

I'm looking at this from the other side. I can't think of a reason that the reverse map should ever retain anything since it is just a convienence and not a cache itself. It should not increment retain counts for anything, that is the responsibility of the cache. It has no need to hold onto sections any longer than they are retained by other objects in the app. 


> As long as the controller is in the cache, I should be able to perform a reverse lookup of the section it's associated with, regardless of whether that specific section object was dereferenced.

Correct - which means as long as the object is held alive by the cache, you can look it up. So why would this object also need to bump the retain count?

> Put in Swift terms, think of controllers as classes & sections as structs. You can't dereference a struct because you can't reference it in the first place. They're "value types," meaning they only represent a specific value, like a number, not a specific object in memory.
  
Not sure this analogy flies or at least I don't follow the implication, we aren't in swift and nothing is a struct.   


If my questions/comments above don't make sense we may need to hangout.